docs/proposal-subscriptions.md

```diff
@@ -0,0 +1,280 @@
+
+Subscriptions are effectively "sets" of data fields, two subscriptions can be "merged" by taking their union. For example the union of the previous two subscriptions is this:
+
+```
+subscription {
+  workflows(ids: [<id>]) {
+    taskProxies {
+      name
+      id
+      status
+      firstParent {
+        id
+      }
+    }
+    familyProxies {
+      id
+      name
+      firstParent {
+      }
+    }
+    nodesEdges {
+      nodes {
+        id
+      }
+      edges {
+        source
+        target
+      }
+    }
+  }
+}
+```
+
+As an added bonus we can query the UI to find out exactly what it is subscribing to and exactly which view has asked for what data making optimisation/debugging much nicer.
+
+### The Workflow Server
+
+When events occur the WS sends "deltas" containing only the changed information.
+
+For example if a job has changed state from submitted to running we might expect a delta of the form:
+
+```
+taskProxies {
+  id: <id>,
+  status: "running"
+  jobs {
+    id: <id>,
+    status: "running"
+  }
+}
+```
+
+### The UI Server
+
+#### Subscriptions (3)
+
+When subscriptions are registered with the UIS they are stored locally in a mapping so we can associate subscriptions with open websockets.
+
+#### Deltas (4)
+
+When deltas arrive from the WS we loop through the subscriptions and take the intersection of the delta and subscription fields, if there is an overlap, this is the delta we send to the UI.
+
+#### Overlaps (5)
+
+For example using the previous delta and subscription:
+
+
+<table>
+  <th>
+    <td>Delta</td>
+    <td>Subscription</td>
+    <td>Intersection</td>
+  </th>
+  <tr>
+    <td>
+        Example 1
+    </td>
+    <td>
+      <pre>
+workflows(ids: [<id>]) {
+  taskProxies {
+    id: <id>,
+    status: "running"
+    jobs {
+      id: <id>,
+      status: "running"
+    }
+  }
+}
+      </pre>
+    </td>
+    <td>
+      <pre>
+subscription {
+  workflows(ids: [<id>]) {
+    taskProxies {
+      name
+      id
+      status
+      firstParent {
+        id
+      }
+    }
+    familyProxies {
+      id
+      name
+      firstParent {
+      }
+    }
+    nodesEdges {
+      nodes {
+        id
+      }
+      edges {
+        source
+        target
+      }
+    }
+  }
+}
+      </pre>
+    </td>
+    <td>
+      <pre>
+workflows(ids: [<id>]) {
+  taskProxies {
+    id: <id>,
+    status: "running"
+  }
+}
+      </pre>
+    </td>
+  </tr>
+</table>
+
+## Caveats
+
+* The comparison of delta and subscription fields will need to be fast.
+  * Else the UIS will become CPU heavy and sluggish.
+  * There may be potential for caching results.
+* GraphQL interface does not provide a mechanism for deltas out of the box.
+  * We may have to extend / implement this ourselves (can we add a `deltaSubscribe` type to GraphQL?).
+  * If we did change the behaviour of the `subscribe` type 3rd party tools (incl graphiql) would loose the ability to work subscriptions (probably not a problem but worth noting).
```

Comment:
I think also `Graphene` is limited to provide always the complete schema requested. So if UI is asking

```
{
a {
  b,
  c
}
```

And only `c` changed, I think Graphene still returns everything, `a`, `b`, and `c`. We may need to either find an option somewhere in Graphene for retrieving only a subset, or modify it (I'm saying Graphene here, but maybe it is done by one of the other graphql dependencies?).

Reply:
Yes, this is the part we would need to solve ourselves.

Reply:
Via Riot chat, @dwsutherland has reputedly figured out how to do this (and without modifying graphene itself).

Reply:
Yes all figured out, PRs tomorrow 🤞